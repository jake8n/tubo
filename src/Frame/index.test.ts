import { Frame } from ".";

describe("Frame", () => {
  let parent: Element, frame: Frame;

  beforeEach(() => {
    parent = document.createElement("div");
    frame = new Frame({ js: "", parent });
  });

  test("is alive", () => {
    expect(frame).toBeTruthy();
  });

  describe("constructor", () => {
    test("append iframe to parent", () => {
      expect(frame.parent.firstChild).toBe(frame.iframe);
    });

    test("perform the first render", () => {
      const spy = jest.spyOn(Frame.prototype, "render");
      frame = new Frame({ js: "", parent });
      expect(spy).toHaveBeenCalled();
    });
  });

  describe("toHTML", () => {
    test("empty state", () => {
      expect(frame.toHTML()).toMatchSnapshot();
    });

    test("custom inputs", () => {
      frame.js = "const message = 'hello'";
      frame.body = '<div id="#app"></div>';
      frame.css = "body { margin: 0; }";
      expect(frame.toHTML()).toMatchSnapshot();
    });
  });

  describe("render", () => {
    test("apply html to iframe", () => {
      frame.iframe = ({
        contentWindow: {
          document: {
            open: jest.fn(),
            close: jest.fn(),
            write: jest.fn(),
          },
        },
      } as unknown) as HTMLIFrameElement;
      frame.render();
      expect(frame.iframe.contentWindow?.document.open).toHaveBeenCalledTimes(
        1
      );
      expect(frame.iframe.contentWindow?.document.write).toHaveBeenCalledWith(
        frame.toHTML()
      );
      expect(frame.iframe.contentWindow?.document.close).toHaveBeenCalledTimes(
        1
      );
    });

    test("re render when new js is applied", () => {
      const spy = jest.spyOn(frame, "render");
      frame.js = "const message = 'hello'";
      expect(spy).toHaveBeenCalled();
    });

    test("re render when new body is applied", () => {
      const spy = jest.spyOn(frame, "render");
      frame.body = '<div id="#app"></div>';
      expect(spy).toHaveBeenCalled();
    });

    test("re render when new css is applied", () => {
      const spy = jest.spyOn(frame, "render");
      frame.css = "body { margin: 0; }";
      expect(spy).toHaveBeenCalled();
    });
  });
});
