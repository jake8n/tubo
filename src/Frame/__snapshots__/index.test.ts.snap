// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Frame toHTML custom inputs 1`] = `
"<head>
  <style>
    body { margin: 0; }
  </style>
</head>
<body>
  <div id=\\"#app\\"></div>
  <script type=\\"module\\">
    const message = 'hello'
  </script>
</body>"
`;

exports[`Frame toHTML empty state 1`] = `
"<head>
  <style>
    body { font-family: sans-serif; }
  </style>
</head>
<body>
  <div id=\\"app\\"></div>
  <script type=\\"module\\">
    
  </script>
</body>"
`;
